lib/graphql_metrics/tracer.rb

```diff
@@ -0,0 +1,127 @@
+# frozen_string_literal: true
+
+module GraphQLMetrics
+  class Tracer
+    # NOTE: Used to store timings from lexing, parsing, validation, before we have a context to store
+    # values in. Uses thread-safe Concurrent::ThreadLocalVar to store a set of values per thread.
+    cattr_accessor :pre_context
+
+    # NOTE: These constants come from the graphql ruby gem.
+    GRAPHQL_GEM_LEXING_KEY = 'lex'
+    GRAPHQL_GEM_PARSING_KEY = 'parse'
+    GRAPHQL_GEM_VALIDATION_KEYS = ['validate', 'analyze_query', 'analyze_multiplex']
+    GRAPHQL_GEM_TRACING_FIELD_KEYS = [
+      GRAPHQL_GEM_TRACING_FIELD_KEY = 'execute_field',
+      GRAPHQL_GEM_TRACING_LAZY_FIELD_KEY = 'execute_field_lazy'
+    ]
+
+    # TODO: can we implement this as an instance method? pass options instead of skip in context?
+    def self.trace(key, data)
+      # NOTE: Context doesn't exist yet during lexing, parsing.
+      possible_context = data[:query]&.context
+
+      skip_tracing = possible_context&.fetch(GraphQLMetrics::SKIP_GRAPHQL_METRICS_ANALYSIS, false)
+      return yield if skip_tracing
+
+      # NOTE: Not all tracing events are handled here, but those that are are handled in this case statement in
+      # chronological order.
+      case key
+      when GRAPHQL_GEM_LEXING_KEY
+        return setup_tracing_before_lexing { yield }
+      when GRAPHQL_GEM_PARSING_KEY
+        return capture_parsing_time { yield }
+      when *GRAPHQL_GEM_VALIDATION_KEYS
+        context = possible_context || data[:multiplex].queries.first.context
```

Comment:
why would we assume that it's always the first query of the multiplex here?

Reply:
Oops. Will fix!

Reply:
We're also not explicitly handling multiplexes in analyzer.rb

I'll write a test to ensure that everything works as expected if we receive a multiplex instead of a single query. That said, if supporting multiplexes becomes complex, we could opt to support them in gem version 3.1.0.

Reply:
Wrote a small summary of why we're OK to ignore multiplexes from the perspective of this gem: https://github.com/Shopify/graphql-metrics/pull/10#issuecomment-555186588